File: examples/typescript/skillcheck.ts
// Security-check a skill or MCP server folder before you install it. Free if someone
// already scanned these exact files; otherwise it pays about $0.03.
//
//   CROSSCHECK_WALLET_KEY=0x... npx tsx skillcheck.ts path/to/skill
import { createHash } from "node:crypto";
import { readdirSync, readFileSync, statSync } from "node:fs";
import { join, relative, sep } from "node:path";
import { postAndWait } from "./pay.js";

const dir = process.argv[2];
if (!dir) throw new Error("Usage: skillcheck.ts <folder>");
const SKIP = new Set(["node_modules", ".git", "dist", "build", "__pycache__", ".venv"]);
const files: Array<{ path: string; content: string }> = [];
const walk = (d: string) => {
  for (const name of readdirSync(d).sort()) {
    const full = join(d, name);
    if (statSync(full).isDirectory()) {
      if (!SKIP.has(name)) walk(full);
    } else {
      const buf = readFileSync(full);
      if (!buf.includes(0) && buf.length <= 200_000) files.push({ path: relative(dir, full).split(sep).join("/"), content: buf.toString("utf8") });
    }
  }
};
walk(dir);

// The bundle hash: SHA-256 of the sorted list of {path, sha256} as canonical JSON.
const sha = (s: string) => createHash("sha256").update(s, "utf8").digest("hex");
const entries = files.map((f) => ({ path: f.path, sha256: sha(f.content) })).sort((a, b) => (a.path < b.path ? -1 : 1));
const bundle = sha(JSON.stringify(entries));

const known = await fetch(`https://crosscheckapi.com/v1/skillcheck/${bundle}`);
const v = known.ok ? await known.json() : (await postAndWait("https://crosscheckapi.com/v1/skillcheck", { files })).verdict;
console.log(`${known.ok ? "(already scanned) " : ""}risk: ${v.risk}. ${v.summary ?? ""}`);
for (const f of v.verdict?.findings ?? v.findings ?? []) console.log(`  ${f.severity.padEnd(8)} ${f.category.padEnd(20)} ${f.file}: ${f.explanation}`);
process.exitCode = v.risk === "critical" || v.risk === "high" ? 1 : 0;

File: examples/python/skillcheck.py
"""Security-check a skill or MCP server folder before you install it. Free if someone already
scanned these exact files; otherwise it pays about $0.03.

    pip install -r requirements.txt
    CROSSCHECK_WALLET_KEY=0x... python skillcheck.py path/to/skill
"""

import asyncio
import hashlib
import json
import os
import sys

import httpx

from pay import post_and_wait

SKIP = {"node_modules", ".git", "dist", "build", "__pycache__", ".venv"}


def read_files(folder: str) -> list[dict]:
    files = []
    for root, dirs, names in os.walk(folder):
        dirs[:] = sorted(d for d in dirs if d not in SKIP)
        for name in sorted(names):
            data = open(os.path.join(root, name), "rb").read()
            if b"\x00" in data or len(data) > 200_000:
                continue  # binary or too large
            path = os.path.relpath(os.path.join(root, name), folder).replace(os.sep, "/")
            files.append({"path": path, "content": data.decode("utf-8")})
    return files


def bundle_sha256(files: list[dict]) -> str:
    """SHA-256 of the canonical JSON list of {path, sha256}, sorted by path."""
    sha = lambda s: hashlib.sha256(s.encode("utf-8")).hexdigest()
    entries = sorted(({"path": f["path"], "sha256": sha(f["content"])} for f in files), key=lambda e: e["path"])
    return sha(json.dumps(entries, separators=(",", ":"), ensure_ascii=False))


async def main(folder: str) -> int:
    files = read_files(folder)
    async with httpx.AsyncClient(timeout=30) as plain:
        known = await plain.get(f"https://crosscheckapi.com/v1/skillcheck/{bundle_sha256(files)}")
    if known.status_code == 200:
        v = known.json().get("verdict") or known.json()
        print("(already scanned)", end=" ")
    else:
        v = (await post_and_wait("https://crosscheckapi.com/v1/skillcheck", {"files": files}))["verdict"]
    print(f"risk: {v['risk']}. {v.get('summary', '')}")
    for f in v.get("findings", []):
        print(f"  {f['severity']:<8} {f['category']:<20} {f['file']}: {f['explanation']}")
    return 1 if v["risk"] in ("critical", "high") else 0


if __name__ == "__main__":
    sys.exit(asyncio.run(main(sys.argv[1])))
